From 79c863697023223cffb0a52b8abd918f35b178cf Mon Sep 17 00:00:00 2001
From: Chanwoo Choi <cw00.choi@samsung.com>
Date: Mon, 9 Sep 2013 19:44:31 +0900
Subject: [PATCH 0801/1302] extcon: Add extcon-port platform device to update
 uevent of cable

This patch add extcon-port platform device to extcon-port driver
to bring up it. extcon-port driver can update uevent of cable
when cable is attached or detached.

Signed-off-by: Chanwoo Choi <cw00.choi@samsung.com>
Signed-off-by: MyungJoo Ham <myungjoo.ham@samsung.com>
---
 drivers/extcon/extcon-port.c | 56 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 56 insertions(+)

diff --git a/drivers/extcon/extcon-port.c b/drivers/extcon/extcon-port.c
index c3121f5..73ce74a 100644
--- a/drivers/extcon/extcon-port.c
+++ b/drivers/extcon/extcon-port.c
@@ -26,6 +26,54 @@
 #include <linux/extcon/extcon-port.h>
 #include <linux/mfd/max77693.h>
 
+/************************************************************/
+/************* extcon-port platform device ******************/
+/************************************************************/
+/*
+ * FIXME: extcon-port driver maintain compatibility of preivous jack device
+ * driver. extcon-port driver have to support DT instead of platform_data way.
+ * Temporarily, add below jack_platform_data to update uevent when cable is
+ * attached or detached.
+ */
+
+/*
+ * extcon-port device
+ */
+#define EXTCON_DEV_MUIC_NAME	"max77693-muic"
+#define EXTCON_DEV_JACK_NAME	"Headset Jack"
+#define EXTCON_DEV_HDMI_NAME	"hdmi"
+
+struct jack_platform_data jack_data = {
+	.usb_online     = 0,
+	.charger_online = 0,
+	.hdmi_online    = 0,
+	.earjack_online = 0,
+	.earkey_online  = 0,
+	.ums_online     = -1,
+	.cdrom_online   = -1,
+	.jig_online     = -1,
+	.host_online    = 0,
+	.cradle_online  = 0,
+
+#ifdef CONFIG_EXTCON_PORT
+	.extcon_name_muic	= EXTCON_DEV_MUIC_NAME,
+	/* .extcon_name_jack	= EXTCON_DEV_JACK_NAME, */
+	/* .extcon_name_hdmi	= EXTCON_DEV_HDMI_NAME, */
+#endif
+};
+
+static struct platform_device extcon_port_device = {
+	.name	= "jack",
+	.id	= -1,
+	.dev	= {
+		.platform_data = &jack_data,
+	},
+};
+
+/************************************************************/
+/************* extcon-port platform driver ******************/
+/************************************************************/
+
 struct extcon_cable_block {
 	char extcon_name[32];
 	char name[32];
@@ -573,6 +621,14 @@ static struct platform_driver extcon_port_driver = {
 
 static int __init extcon_port_init(void)
 {
+	int ret;
+
+	ret = platform_device_register(&extcon_port_device);
+	if (ret < 0) {
+		pr_err("Failed to register extcon-port device\n");
+		return ret;
+	}
+
 	return platform_driver_register(&extcon_port_driver);
 }
 late_initcall(extcon_port_init);
-- 
1.8.3.2

